events/events.go

```diff
@@ -0,0 +1,205 @@
+	emitUserEvent(
+		help,
+		false,
+		&pb.Event_PoetWaitChallenge{
+			PoetWaitChallenge: &pb.EventPoetWaitChallenge{
+				Publish: publish.Uint32(),
+				Target:  target.Uint32(),
+				Wait:    durationpb.New(wait),
+			},
+		},
+	)
+}
+
+func EmitPostStart(challenge []byte) {
+	const help = "Node started post execution for the challenge from poet."
+	emitUserEvent(
+		help,
+		false,
+		&pb.Event_PostStart{PostStart: &pb.EventPostStart{Challenge: challenge}},
+	)
+}
+
+func EmitPostComplete(challenge []byte) {
+	const help = "Node finished post execution for challenge."
+	emitUserEvent(
+		help,
+		false,
+		&pb.Event_PostComplete{PostComplete: &pb.EventPostComplete{Challenge: challenge}},
+	)
+}
+
+func EmitPostFailure() {
+	const help = "Node failed post execution."
+	emitUserEvent(
+		help,
+		true,
+		&pb.Event_PostComplete{PostComplete: &pb.EventPostComplete{}},
+	)
+}
+
+func EmitAtxPublished(
+	current, target types.EpochID,
+	id types.ATXID,
+	wait time.Duration,
+) {
+	const help = "Published activation for the current epoch. " +
+		"Node needs to wait till the start of the target epoch in order to be eligible for rewards."
+	emitUserEvent(
+		help,
+		false,
+		&pb.Event_AtxPublished{
+			AtxPublished: &pb.EventAtxPubished{
+				Current: current.Uint32(),
+				Target:  target.Uint32(),
+				Id:      id[:],
+				Wait:    durationpb.New(wait),
+			},
+		},
+	)
+}
+
+func EmitEligibilities(
+	epoch types.EpochID,
+	beacon types.Beacon,
+	atx types.ATXID,
+	activeSetSize uint32,
+	eligibilities map[types.LayerID][]types.VotingEligibility,
+) {
+	const help = "Computed eligibilities for the epoch. " +
+		"Rewards will be received after publishing proposals at specified layers. " +
+		"Total amount of rewards in SMH will be based on other participants in the layer."
+	emitUserEvent(
+		help,
+		false,
+		&pb.Event_Eligibilities{
+			Eligibilities: &pb.EventEligibilities{
+				Epoch:         epoch.Uint32(),
+				Beacon:        beacon[:],
+				Atx:           atx[:],
+				ActiveSetSize: activeSetSize,
+				Eligibilities: castEligibilities(eligibilities),
+			},
+		},
+	)
+}
+
+func castEligibilities(proofs map[types.LayerID][]types.VotingEligibility) []*pb.ProposalEligibility {
+	rst := make([]*pb.ProposalEligibility, 0, len(proofs))
+	for lid, eligs := range proofs {
+		rst = append(rst, &pb.ProposalEligibility{
+			Layer: lid.Uint32(),
+			Count: uint32(len(eligs)),
+		})
+	}
+	return rst
+}
+
+func EmitProposal(layer types.LayerID, proposal types.ProposalID) {
+	const help = "Published proposal. Rewards will be received, once proposal is included into the block."
+	emitUserEvent(
+		help,
+		false,
+		&pb.Event_Proposal{
+			Proposal: &pb.EventProposal{
+				Layer:    layer.Uint32(),
+				Proposal: proposal[:],
+			},
+		},
+	)
+}
+
+func emitUserEvent(help string, failure bool, details pb.IsEventDetails) {
+	mu.RLock()
+	defer mu.RUnlock()
+	if reporter != nil {
```

Comment:
I think inversing the if and returning early makes the code clearer.

Any reason that `emitUserEvent` is a function instead of a method on `reporter`? It's a bit odd that the reporter has to be initialized with `InitializeReporter` and every function in the package needs to check if it is `nil` or not.

If it was a service and we don't want events to be reported we can just pass a `NoOpEventReporter` that implements the same interface as the actual reporter.

Probably not for this PR but technical debt that we should clean up in the future.

Reply:
I'm also in favour of passing the reporter to whatever services need to report. I made a ticket for this [here](https://github.com/spacemeshos/go-spacemesh/issues/4651)

Reply:
yes it should not be global, thats legacy.. which i am extending ofcourse, but can't fix it right now